pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -231,91 +229,220 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+		return reconcile.Result{}, fmt.Errorf("can not retrieve kubeone manifest secret: %w", err)
+	}
+	currentManifest := manifestSecret.Data[resources.KubeOneManifest]
+
+	_, err = r.upgradeAction(ctx, log, externalCluster, externalClusterProvider, currentManifest, clusterStatus)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err = r.checkPodStatusIfExists(ctx, log, externalCluster, UpgradeControlPlaneAction); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	_, err = r.migrateAction(ctx, log, externalCluster, externalClusterProvider, currentManifest, clusterStatus)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err = r.checkPodStatusIfExists(ctx, log, externalCluster, MigrateControlPlaneAction); err != nil {
+		return reconcile.Result{}, err
+	}
 
 	return reconcile.Result{}, nil
 }
 
-func (r *reconciler) checkPodStatus(ctx context.Context, log *zap.SugaredLogger, pod *corev1.Pod, externalCluster *kubermaticv1.ExternalCluster) error {
+func returnWantedContainerRuntime(currentManifest []byte) (string, error) {
+	cluster := &kubeonev1beta2.KubeOneCluster{}
+	if err := yaml.UnmarshalStrict(currentManifest, cluster); err != nil {
+		return "", fmt.Errorf("failed to decode manifest secret data: %w", err)
+	}
+	if cluster.ContainerRuntime.Docker != nil {
+		return resources.ContainerRuntimeDocker, nil
+	} else if cluster.ContainerRuntime.Containerd != nil {
+		return resources.ContainerRuntimeContainerd, nil
+	}
+	return "", nil
+}
+
+func (r *reconciler) checkPodStatus(ctx context.Context,
+	log *zap.SugaredLogger,
+	pod *corev1.Pod,
+	externalCluster *kubermaticv1.ExternalCluster,
+	action string,
+) error {
 	log.Debugw("Checking kubeone pod status", "Pod", pod.Name)
-	for pod.Status.Phase == corev1.PodSucceeded {
+	if pod.Status.Phase == corev1.PodSucceeded {
 		oldexternalCluster := externalCluster.DeepCopy()
 		// update kubeone externalcluster status.
 		externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status = kubermaticv1.StatusRunning
 
 		if err := r.Patch(ctx, externalCluster, ctrlruntimeclient.MergeFrom(oldexternalCluster)); err != nil {
-			return fmt.Errorf("failed to add kubeconfig reference to %s: %w", externalCluster.Name, err)
+			return fmt.Errorf("failed to update cluster status %s: %w", externalCluster.Name, err)
 		}
-		log.Debug("KubeOne Cluster Upgraded!")
-		if err := r.Delete(ctx, pod); err != nil {
-			return err
+		if action == UpgradeControlPlaneAction {
+			log.Debug("KubeOne Cluster Upgraded!")
+		} else if action == MigrateControlPlaneAction {
+			log.Debug("KubeOne Cluster Migrated!")
 		}
-	}
-	if pod.Status.Phase == corev1.PodFailed {
-		upgradeErr := fmt.Sprintf("Failed to upgrade kubeone cluster %s", externalCluster.Name)
+	} else if pod.Status.Phase == corev1.PodFailed {
+		upgradeErr := fmt.Sprintf("Failed to %s kubeone cluster %s", action, externalCluster.Name)
```

Comment:
I don't think we need to include the cluster name in this error message. It's being used in the ExternalCluster object itself (so the context is clear) and it's being logged (and I think the logger already contains the cluster name as well).